### ISSUE TITLE
Add PredeclaredClassAssignedToVariable inspection

### ISSUE THREAD

**Issue by pchemguy** [rubberduck-vba/Rubberduck#5876]
**What**
Potential bug: predeclared class instance is assigned to a variable.
Potential reason: missed *New* operator.
Automatic fix: Add missing *New* operator.

**Why**
Predeclared class instances are typically used as class factories. Usually, assigning the predeclared instance to a variable would make little sense.

**Example**
This code should trigger the inspection:

```vb
'@PredeclaredId

Public Function Create(ByVal Arg As String) As PredeclaredClass
    Dim Instance As PredeclaredClass
    Set Instance = PredeclaredClass
    Instance.Init Arg
    Set Create = Instance
End Function
```

---

**QuickFixes**
Should Rubberduck offer one or more quickfix(es) for this inspection? Describe them here (note: all inspections allow for `IgnoreOnceQuickFix`, unless explicitly specified):

1. **Add missing New operator**

    Example code, after quickfix is applied:

```vb
'@PredeclaredId

Public Function Create(ByVal Arg As String) As PredeclaredClass
    Dim Instance As PredeclaredClass
    Set Instance = New PredeclaredClass
    Instance.Init Arg
    Set Create = Instance
End Function
```

---

**Resources**
Each inspection needs a number of resource strings - please provide a suggestion here:

 - **Predeclared class instance is assigned to variable**: the name of the inspection, as it appears in the inspection settings.
 - **Predeclared class instance is usually used as a class factory, and it should make little sense to assign it to a variable.**: the detailed rationale for the inspection, as it appears in the inspection results toolwindow's bottom panel.
 - **_Predeclared class {0} is assigned to variable_, {0}:the name of the predeclared class**: the resource string for an inspection result; if the string needs placeholders, identify what goes in for them (e.g. `Variable {0} is not used`, {0}:the name of the variable).

**Comment by FullValueRider** [comment]
Personally I'd write

```
set Instance = PredeclaredClass.Init(arg)
```

And consequently I'd suggest that your argument above needs to be reversed so that we get an inspection warning if we use New with a PredeclaredId. 

 e.g.
```
set Instance = New PredeclaredClass

```
should trigger the inspection warning.

**Comment by Vogel612** [comment]
@FullValueRider I like that idea, but that honestly sounds like a separate inspection that is contingent on having such an `Init` method (or more generally a method on the Predeclared class that returns a `new` instance of the class itself)

**Comment by pchemguy** [comment]
> Personally I'd write
> 
> ```
> set Instance = PredeclaredClass.Init(arg)
> 
> should trigger the inspection warning.

This is wrong! Init is a constructor! Factory is responsible for New'ing the class and is called on the predeclared instance. Constructor is responsible for initialization and must be called on the new instance. That is the reason why we need a factory and constructor. One of them is not sufficient.

**Comment by FullValueRider** [comment]
I suspect the methodology was used to avoid creating an instance variable.  I just copied what I thought was a good idea.  Nothing wrong with your version though.

**Comment by Energeer** [comment]
Came across this today. Is there any sort of convention/example for a guard clause that can be written for when a predeclared (factory) class is used as a variable? Particularly for when the factory creates the class as an interface and the variable is the same interface.

I was thinking like having a Boolean member variable in the class which is only true if the factory method is used?

Thanks!
